### PR TITLE
skip blocks with out-of-order chunk during compaction

### PR DESCRIPTION
**Pull request by huyan0**
Signed-off-by: Yang Hu <yhuz@amazon.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
- add a check for OOO chunks in block index, and mark such block for no compaction, and continue on with the compaction
- add testutil to generate an index with OOO chunks

## Verification
- verified the exepected behavior with `compact_e2e_test.go`

Addresses: #3442 
cc: @yeya24 @alvinlin123 @roystchiang

### CODE REVIEW

File: cmd/thanos/compact.go
Comment (yeya24) [inline cmd/thanos/compact.go:352]: I am wondering if we need a new `reason` label for this mark counter. Now we cannot differentiate any compact blocks due to large index or out of order chunks.

**Review by yeya24 — COMMENTED** [review]
Love it! Super high-quality code. Thanks for the contribution @huyan0

File: pkg/block/index.go
Comment (alvinlin123) [inline pkg/block/index.go:114]: I am wondering, do you think we can just skip compaction for all the CriticalErr; instead of just OutOfOrderChunksErr?

Comment (huyan0) [reply pkg/block/index.go:114]: I think it's okay on the compact side given we cannot generate more blocks than not compacting...I don't think there's much implication on the retrieval side either but would like to hear what @yeya24 thinks : )
Could you also provide some context on why the project decided to halt compaction in the first place?

File: cmd/thanos/compact.go
Comment (huyan0) [reply cmd/thanos/compact.go:352]: sounds reasonable, especially if we want to skip all blocks with critical error as @alvinlin123 suggested

**Comment by huyan0** [comment]
I have updated the metric counter, will add to changelog if there's no further comments for now

**Review by yeya24 — DISMISSED** [review]
This pr LGTM. But it changes the default behavior and might be critical so it would be good to hear from other maintainers.

cc @thanos-io/thanos-maintainers

**Comment by wiardvanrij** [comment]
The code looks solid and you referenced an issue. If I understand correctly, we now mark it as corrupted and skip it. That still leaves a block corrupted. Are there any follow-up steps that a user could do? Basically I would really love to have some documentation on:
- How could one find corrupted blocks?
- What is the impact for the user if that happens?
- Could one still fix a corrupted block and let it get compacted after the fact?
- anything else that provides context to users

This would IMO be the cherry on top :)

**Comment by bwplotka** [comment]
TODO: Can we check if those blocks are *readable* by Thanos Store/Cortex Store

**Comment by bwplotka** [comment]
TODO: Review & check for consistency with other issues and how we handle them (:

**Comment by huyan0** [comment]
@bwplotka  regarding your two comments:

1. the blocks are still readable because the mark that we add to the block is for compact only, and the store API has no requirement on whether the chunks are sorted: [link](https://github.com/thanos-io/thanos/blob/aa148f8fdb281a9ea4c1e5a7ee2e1336d8459c4f/pkg/store/storepb/rpc.pb.go#L703)

2. We handle this error differently than the other errors encountered during compaction; for those we halt the compaction. Other than OOO chunks, the two other specific errors compact check for [are these](https://github.com/thanos-io/thanos/blob/aa148f8fdb281a9ea4c1e5a7ee2e1336d8459c4f/pkg/compact/compact.go#L752) , for which we attempt to repair or compacts if a feature flag is set, so I don't see an overarching consistent approach to adhere to, other than logging the occurrence to provide context for user. This is understandable since these errors are unexpected and specific. I agree a deep dive into the root cause is needed to address this in long term to come up with a more coherent fix.

File: pkg/block/index.go
Comment (huyan0) [reply pkg/block/index.go:114]: @bwplotka  do you have some context on the discussion? :)

**Comment by huyan0** [comment]
PR still pending review :) Any updates needed?

File: cmd/thanos/compact.go
Comment (yeya24) [inline cmd/thanos/compact.go:674]: What about `compact.skip-block-with-out-of-order-chunks`?

**Comment by yeya24** [comment]
Btw let's update the changelog and we are ready to go I think

**Comment by yeya24** [comment]
@huyan0 Sorry for the late review. I will merge the pr after you resolve the conflict.

**Review by yeya24 — APPROVED** [review]
LGTM

File: cmd/thanos/compact.go
Comment (yeya24) [inline cmd/thanos/compact.go:151]: Let's add other reasons for no compact marker.

File: cmd/thanos/compact.go
Comment (yeya24) [inline cmd/thanos/compact.go:152]: Panic here. We need a reason for deletion marker. Empty string is fine here.

File: cmd/thanos/compact.go
Comment (yeya24) [inline cmd/thanos/compact.go:350]: Panic again here.

File: pkg/compact/compact.go
Comment (yeya24) [inline pkg/compact/compact.go:981]: typo `copmaction` and please add a period here

**Review by yeya24 — CHANGES_REQUESTED** [review]
There might be other panics caused by that deletion marker metric. Please check and fix them. Thanks for the patience.

**Comment by yeya24** [comment]
Can you do a rebase on latest main rather than including changes from another pr.

File: CHANGELOG.md
Comment (yeya24) [inline CHANGELOG.md:20]: The pr number seems incorrect in the link.

File: pkg/compact/compact.go
Comment (yeya24) [inline pkg/compact/compact.go:561]: Let's remove this line

File: pkg/testutil/testutil.go
Comment (yeya24) [inline pkg/testutil/testutil.go:224]: uppercase

File: pkg/testutil/testutil.go
Comment (yeya24) [inline pkg/testutil/testutil.go:315]: return iw.Close()

**Review by yeya24 — APPROVED** [review]
LGTM

**Comment by huyan0** [comment]
looking... Sorry for the back and forth. Will try to address it

**Review by yeya24 — APPROVED** [review]
LGTM! Thanks for the contribution.

**Pull request merged**
